### PR TITLE
Allow to connect with gocloud to AWS/GCP instances.

### LINKED ISSUE #21: Connect to cloud instances natively
Hi,

in the previous repository there was an open feature request https://github.com/hashicorp/terraform-provider-postgresql/issues/110 to allow to natively connect to cloud instances like Cloud SQL Proxy for GCP and AWS/Azure equivalent.

@benfdking implemented that feature in their fork here: https://github.com/tumelohq/terraform-provider-postgresql/ using https://gocloud.dev/.

Would you consider such approach if I were to submit a patch? I am waiting on https://github.com/cyrilgdn/terraform-provider-postgresql/pull/5 first before I start working on it.

# References
https://github.com/benfdking/terraform-provider-postgresql/pull/1
https://gocloud.dev/howto/sql/

### PR DESCRIPTION
**Pull request by cyrilgdn**
[GoCloud](https://gocloud.dev/howto/sql/) simplifies connection to AWS/GCP Postgresql instances.
It manages TLS certificates, GCP authentication proxy, etc...

This PRs defines a new `scheme` settings to allow switching to GoCloud instead of [lib/pq](https://github.com/lib/pq).

See https://github.com/cyrilgdn/terraform-provider-postgresql/blob/gocloud/website/docs/index.html.markdown#gocloud

Depends on #28 
Vendors update in #31 

Fix #21

### CODE REVIEW

File: website/docs/index.html.markdown [hunk 1]
```diff
@@ -105,3 +110,55 @@ The following arguments are supported:
   Version](https://www.postgresql.org/support/versioning/) or `current`.  Once a
   connection has been established, Terraform will fingerprint the actual
   version.  Default: `9.0.0`.
+
+## GoCloud
+
+By default, the provider uses [lib/pq][libpq] library to connect to PostgreSQL servers, but you can switch to [GoCloud](https://gocloud.dev/howto/sql/).
+GoCloud simplifies the connection to AWS/GCP hosted databases. For example GCP requires a custom proxy with authentication credentials, GoCloud manages it automatically.
```
Comment (benfdking) [inline website/docs/index.html.markdown:117]:
```suggestion
By default, the provider uses the [lib/pq][libpq] library to directly connect to PostgreSQL host instance. For connections to AWS/GCP hosted instances, the provider can connect through the [GoCloud](https://gocloud.dev/howto/sql/) library. GoCloud simplifies connecting to AWS/GCP hosted databases, managing any proxy or custom authentication details.
```

File: website/docs/index.html.markdown [hunk 2]
```diff
@@ -105,3 +110,55 @@ The following arguments are supported:
...
+
+### AWS
+
+To enable GoCloud for AWS RDS, set `scheme` to `awspostgres`, `host` is the database endpoint in RDS.
```
Comment (benfdking) [inline website/docs/index.html.markdown:121]:
```suggestion
To enable GoCloud based connections to AWS RDS instances, set `scheme` to `awspostgres` and `host` to the RDS database's endpoint value.
```

File: website/docs/index.html.markdown [hunk 3]
```diff
@@ -105,3 +110,55 @@ The following arguments are supported:
...
+(e.g.: `instance.xxxxxx.region.rds.amazonaws.com`)
+
```
Comment (benfdking) [inline website/docs/index.html.markdown:123]: Although I haven't tried it myself, similarly to Google I presume you need to setup credentials for RDS connections. My understanding is that like GCP, it uses a Platform (GCP/AWS) with the right permissions to create that connection tunnel.

File: website/docs/index.html.markdown [hunk 4]
```diff
@@ -105,3 +110,55 @@ The following arguments are supported:
...
+  port     = 5432
+  password = "test1234"
+
+  superuser = false
+}
+```
+
+### GCP
+
+To enable GoCloud for GCP SQL, set `scheme` to `gcppostgres`, host is the connection name in following format: `project/region/instance`
```
Comment (benfdking) [inline website/docs/index.html.markdown:138]:
```suggestion
To enable GoCloud for GCP SQL, set `scheme` to `gcppostgres` and `host` to the connection name of the instance in following format: `project/region/instance`.
```

**Review by benfdking — COMMENTED** [review]
Hey amazing work! It's much better documented and much more intuitive to use with the `scheme` setup. It all looks good to me. Thank you so much for taking on the task of maintaining this project! 🙏  I do hope the little documentation comments are helpful.

File: website/docs/index.html.markdown (hunk 3 above)
Comment (cyrilgdn) [reply website/docs/index.html.markdown:122]: AWS does not have a mandatory proxy like GCP, so with public accessible instance this example work. Of course most user will probably need to use a tunnel but I'm not sure if Gocloud could help for that.

**Review by cyrilgdn — COMMENTED** [review]
Thanks for you review!

**Comment by ademariag** [comment]
Moving conversation here:

@cyrilgdn perhaps I am not getting something, but before I could use [google_sql_database_instance.antha-database.connection_name](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_database_instance#connection_name )
 which gives the `project:region:instance` format. However now if you expect the `project/region/instance` format, how do I easily link from an already existing instance?
 
 I think it would be nice if it could seamlessly handle both `project:region:instance` and  `project/region/instance`

**Comment by NickyMateev** [comment]
Hi guys - great job introducing this! 🎉

I'm interested in how this PR progresses and eventually goes forward as I myself have a use case of my own which requires similar functionalities, but I'm not quite sure if this would enable it.

Essentially I'm dealing with a terraform infrastructure which sets up a GCP project and one of the resources it sets up is a CloudSQL instance and a bunch of admin users with the `google_sql_database` and `google_sql_user`. Unfortunately the `google_sql_user` primitive does not allow configuring the psql user to be readonly (have permission only to do SELECT statements). So I looked in the direction of using the postgresql terraform provider and it's primitives directly to place the necessary permissions(restrictions) to the read-only psql users, but then I hit the problem with GCP and the CloudSQL proxy it requires.

This PR looks to solve this issue, but I'm still wondering if it would work as from [the documentation](https://github.com/cyrilgdn/terraform-provider-postgresql/blob/gocloud/website/docs/index.html.markdown#gocloud) it seems like you have to provide psql username and password to work with when using the postgresql provider primitives. But what if the terraform installation has not created the admin user with the google cloud sql primitives so that the postgresql provider could use it. Maybe this is a general terraform question about ordering execution, but was still wondering if you could give me a hint with this.

Thanks!

**Comment by ademariag** [comment]
Hi Nick, I use a similar setup.

I create the "`admin`" user using `google_sql_user`, and then I reference that user/password from the postgres provider.

Key to this is the "Lazy Initialisation" that @cyrilgdn merged (https://github.com/cyrilgdn/terraform-provider-postgresql/pull/5) a couple of days ago, you should be able to configure dependencies so that the provider is only initialised after the `google_sql_user` user has been provisioned.

**Comment by NickyMateev** [comment]
Hi @ademariag ,

Thanks for the quick response!

Do you know if the `1.11.0-pre-gocloud` version [here](https://registry.terraform.io/providers/cyrilgdn/postgresql/latest) contains this lazy initialization? I'd like to test if this approach would fit my use case.

**Comment by ademariag** [comment]
yes it does

**Comment by cyrilgdn** [comment]
> I'm interested in how this PR progresses 

Hi @NickyMateev , thanks for your interest on that. we'll probably finish this next week-end.

> This PR looks to solve this issue, but I'm still wondering if it would work for my use case as from the documentation

We could upgrade the documentation to have a full example with the instance creation indeed, could be a good idea. (same for AWS)

**Comment by cyrilgdn** [comment]
> Moving conversation here:
> 
> @cyrilgdn perhaps I am not getting something, but before I could use [google_sql_database_instance.antha-database.connection_name](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_database_instance#connection_name)
> which gives the `project:region:instance` format. However now if you expect the `project/region/instance` format, how do I easily link from an already existing instance?
> 
> I think it would be nice if it could seamlessly handle both `project:region:instance` and `project/region/instance`

@ademariag I removed it on purpose to follow the Gocloud documentation format and I thought it was not a big deal for the user to use the right format, but I didn't thought about this use case.

I'll add back the support of both `project:region:instance` and `project/region/instance`

**Comment by ademariag** [comment]
thank you @cyrilgdn much appreciated

**Comment by cyrilgdn** [comment]
@ademariag @benfdking @NickyMateev 

FYI, this has just been released in [v1.11.0](https://registry.terraform.io/providers/cyrilgdn/postgresql/latest).

**Comment by ademariag** [comment]
this is so cool. thank you very much!

**Comment by benfdking** [comment]
Amazing work!

**Pull request merged**